src/error/error.h

```diff
@@ -0,0 +1,158 @@
+#pragma once
+
+#include <string>
+
+// Error utilities.
+
+// Base error class, with a message.
+// All error classes should derive from this one.
+class Error {
+ public:
+  explicit Error(const std::string& message) : message_(message) {}
+
+  virtual std::string toString() const { return message(); }
+
+ protected:
+  std::string message() const { return message_; }
+ private:
+  const std::string message_;
+};
+
+
+// Either an error or a T.
+// This class is immutable (but the T can be mutable).
+template <typename T, typename Err = Error>
+class ErrorOr {
+ public:
+  // Not explicit on purpose.
+  ErrorOr(const Err& error) : is_ok_(false) {
+    // Placement new.
+    new (&union_.error) Err(error);
+  }
+
+  // Not explicit on purpose.
+  ErrorOr(Err&& error) : is_ok_(false) {
+    // Placement new.
+    new (&union_.error) Err(std::move(error));
+  }
+
+  // Not explicit on purpose.
+  ErrorOr(const T& value) : is_ok_(true) {
+    // Placement new.
+    new (&union_.value) T(value);
+  }
+
+  // Not explicit on purpose.
+  ErrorOr(T&& value) : is_ok_(true) {
+    // Placement new.
+    new (&union_.value) T(std::move(value));
+  }
+
+  ErrorOr(const ErrorOr& other) : is_ok_(other.is_ok_) {
+    if (is_ok_)
+      union_.value = other.union_.value;
+    else
+      union_.error = other.union_.error;
+  }
+
+  ErrorOr(ErrorOr&& other) : is_ok_(other.is_ok_) {
+    if (is_ok_)
+      new (&union_.value) T(std::move(other.union_.value));
+    else
+      new (&union_.error) Err(std::move(other.union_.error));
+  }
+
+  ErrorOr& operator=(const ErrorOr& other) {
+    ~ErrorOr();
+    is_ok_ = other.is_ok_;
+    if (is_ok_)
+      union_.value = other.union_.value;
+    else
+      union_.error = other.union_.error;
+  }
+
+  ErrorOr& operator=(ErrorOr&& other) {
```

Comment:
If the current object contains a value and the argument an error (or the opposite), this will not work. The safest way is to delete the current object and then move the right member into the union (like in the move constructor).